### PR TITLE
handle frame in stack trace that doesn't have a module

### LINKED ISSUE #1739: [Bug] failure while sanitiainzg exception stack trace in some cases
In some scenarios, one of the insepected frames in the stack trace does not have a module, resulting in a failure to print the underlying error.

The following looks like it's resolving it:

```diff
++ hydra/_internal/utils.py
@@ -248,6 +248,8 @@
                 while end is not None:
                     frame = end.tb_frame
                     mdl = inspect.getmodule(frame)
+                    if mdl is None:
+                        break
-                     assert mdl is not None
```

### PR DESCRIPTION
**Pull request by Jasha10**
Closes #1739

Not sure how to test this.
Currently there are not any unit tests of the `run_and_report` function.

### CODE REVIEW

**Comment by omry** [comment]
Yeah, it's a hard one to test.

I am actually not sure that the best handling is to break.
Maybe we should revert to printing the full stacktrace in such a case?

**Comment by omry** [comment]
you can test by triggering the code path to see what happens.

**Comment by omry** [comment]
I would say that in any case there is any exception at all while trying print a nice exception in that branch we should revert to printing the full exception.

**Comment by Jasha10** [comment]
I wrapped the nice-exception logic in a try/except block.

File: hydra/_internal/utils.py [hunk 1]
```diff
@@ -213,75 +213,79 @@ def run_and_report(func: Any) -> Any:
...
+                        cur = cur.tb_next
+                        assert iter_tb.tb_next is not None
+                        iter_tb = iter_tb.tb_next
+
+                    print_exception(etype=None, value=ex, tb=final_tb)  # type: ignore
+                sys.stderr.write(
+                    "\nSet the environment variable HYDRA_FULL_ERROR=1 for a complete stack trace.\n"
+                )
+            except Exception:
+                print_exception(etype=None, value=ex, tb=ex.__traceback__)
```
Comment (omry) [inline hydra/_internal/utils.py:288]: Probably better to raise the original exception instead after printing some kind of warning.
```suggestion
                # print something to stderr here.
                raise ex
```

**Comment by omry** [comment]
Can you check the CI failures?

File: hydra/_internal/utils.py (hunk 1 above)
Comment (Jasha10) [reply hydra/_internal/utils.py]: Ok, as of commit https://github.com/facebookresearch/hydra/pull/1741/commits/7de5282b18c33f511d4c15f174120c0ea0ce6cb5 a warning is printed including the filename / line number where the second exception occurred.

**Comment by Jasha10** [comment]
> Can you check the CI failures?

Ok, I've fixed most of the failures.
There are still a few ray_launcher failures, but I think they're unrelated.

File: hydra/_internal/utils.py (hunk 1 above)
Comment (omry) [reply hydra/_internal/utils.py]: The second exception is happening only in the processing logic.
It seems like an overkill.

You can simplify to something like:
```
                sys.stderr.write(
                    "An error occurred during Hydra's exception formatting:"
                    + os.linesep + repr(ex2)
                )
                raise ex
```

Comment (Jasha10) [reply hydra/_internal/utils.py]: Done in https://github.com/facebookresearch/hydra/pull/1741/commits/cf7bbbfcfeab1d754fb9823f1292033c12004150.

File: tests/test_utils.py [hunk 2]
```diff
@@ -60,3 +65,120 @@ def test_to_absolute_path_without_hydra(
     path = str(Path(path))
     expected = str(Path(expected).absolute())
     assert utils.to_absolute_path(path) == expected
+
+
+class TestRunAndReport:
+    class DemoFunctions:
+        """Demo inputs for `run_and_report`"""
```
Comment (omry) [inline tests/test_utils.py:72]: Can you explain the testing logic?
The test is surprisingly large.

Comment (Jasha10) [reply tests/test_utils.py:72]: > The test is surprisingly large.

Sorry about that.
This is not just testing the new functionality. It is also testing the pre-existing `run_and_report` behavior.

The `run_and_report` function has the following signature:
```
def run_and_report(func: Any) -> Any: ...
```

There are three test methods:
- `test_success` tests when `run_and_report` runs `func` successfully.
- `test_failure` tests when `func` raises an exception and `run_and_report` is able to print a nicely-formatted error message.
- `test_simplified_traceback_failure` tests when `func` raises an exception and `run_and_report` is *not* able to print a nicely-formatted error message -- this is the case where the original exception gets re-raised.

There are four `@staticmethod` definitions. These are passed in to `run_and_report` as the `func` argument. They are designed to trigger different logic code paths inside of `run_and_report`:
- `success_func` raises no exception
- `simple_error` raises an `AssertionError`
- `run_job_wrapper` triggers the special logic in `run_and_report` that looks for a function called `"run_job"` in the stack, and strips away the leading stack frames
- `omegaconf_job_wrapper` triggers the special logic in `run_and_report` that looks for the `omegaconf` module in the stack, and strips away the bottom stack frames

Comment (omry) [reply tests/test_utils.py:72]: Gotcha.
Thanks for adding this. Can you add a comment describing this in the test class?
(does not need to be as detailed).

**Review by omry — APPROVED** [review]
Overall looks good.
See comment.

**Comment by omry** [comment]
Looking good, don't forget a bugfix news fragment.

**Pull request merged**